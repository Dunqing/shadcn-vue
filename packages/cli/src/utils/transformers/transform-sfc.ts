import { createRequire } from 'node:module'
import type { Config } from '../get-config'
import { transformImport } from './transform-import'

const require = createRequire(import.meta.url)
const { transform } = require('detype')

export async function transformByDetype(content: string, filename: string) {
  return await transform(content, filename, {
    removeTsComments: true,
  })
}

interface File {
  name: string
  content: string
}

export async function transformSFC(file: File, config: Config) {
  let content = transformImport(file.content, config)
  if (config.typescript)
    content = await transformByDetype(content, file.name)

  return content
}
