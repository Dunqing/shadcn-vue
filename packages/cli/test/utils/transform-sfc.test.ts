import { describe, expect, test } from 'vitest'
import { transform } from '../../src/utils/transformers'

describe('transformSFC', () => {
  test('basic', async () => {
    const result = await transform({
      filename: 'app.vue',
      raw: `<script lang="ts" setup>
      const array: (number | string)[] = [1, 2, 3]
      </script>
      
      <template>
        <div v-bind="{ array }">
          template
        </div>
      </template>
      
      <style scoped>
      </style>
      `,
      config: {},
    })
    expect(result).toMatchSnapshot()
  })

  test('defineProps', async () => {
    const result = await transform({
      filename: 'app.vue',
      raw: `<script lang="ts" setup>
      const props = defineProps<{ foo: string }>()
      </script>
      `,
      config: {},
    })
    expect(result).toMatchSnapshot()
  })

  test('defineProps with withDefaults', async () => {
    const result = await transform({
      filename: 'app.vue',
      raw: `<script lang="ts" setup>
      const props = withDefaults(defineProps<{ foo: string }>(), {
        foo: 'bar'
      })
      </script>
      `,
      config: {},
    })
    expect(result).toMatchSnapshot()
  })

  test('defineEmits', async () => {
    const result = await transform({
      filename: 'app.vue',
      raw: `<script lang="ts" setup>
      const emit = defineEmits<{ foo: string }>()
      </script>
      `,
      config: {},
    })
    expect(result).toMatchSnapshot()
  })
})
