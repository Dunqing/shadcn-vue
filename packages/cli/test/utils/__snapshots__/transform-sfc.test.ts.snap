// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transformSFC > basic 1`] = `
"<script setup>
const array = [1, 2, 3];
</script>

<template>
  <div v-bind=\\"{ array }\\">template</div>
</template>

<style scoped></style>
"
`;

exports[`transformSFC > defineEmits 1`] = `
"<script setup>
const emit = defineEmits([\\"foo\\"]);
</script>
"
`;

exports[`transformSFC > defineProps 1`] = `
"<script setup>
const props = defineProps({
  foo: { type: String, required: true },
});
</script>
"
`;
